packages/io-messages-common/src/adapters/lollipop/lollipop-middleware.ts

```diff
@@ -0,0 +1,139 @@
+import { HttpRequest, InvocationContext } from "@azure/functions";
+import { ulid } from "ulid";
+
+import { userIdentitySchema } from "../auth/user-identity.js";
+import { Middleware, MiddlewareError } from "../middleware.js";
+import {
+  assertionRefSha256Schema,
+  assertionRefSha384Schema,
+  assertionRefSha512Schema,
+} from "./definitions/assertion-ref.js";
+import {
+  LollipopHeaders,
+  lollipopHeadersSchema,
+} from "./definitions/lollipop-headers.js";
+import {
+  JwkPubKeyHashAlgorithm,
+  jwkPubKeyHashAlgorithmSchema,
+} from "./definitions/pub-key-algorithm.js";
+import { lollipopRequestHeadersSchema } from "./definitions/request-headers.js";
+import { LollipopSignatureInput } from "./definitions/signature-input.js";
+import { Thumbprint, thumbprintSchema } from "./definitions/thumbprint.js";
+import LollipopClient, { LollipopClientError } from "./lollipop-client.js";
+
+export const lollipopExtraInputsCtxKey = "lollipopHeaders";
+
+const algoSchemaMap: {
+  algo: JwkPubKeyHashAlgorithm;
+  schema:
+    | typeof assertionRefSha256Schema
+    | typeof assertionRefSha384Schema
+    | typeof assertionRefSha512Schema;
+}[] = [
+  {
+    algo: jwkPubKeyHashAlgorithmSchema.Enum.sha256,
+    schema: assertionRefSha256Schema,
+  },
+  {
+    algo: jwkPubKeyHashAlgorithmSchema.Enum.sha384,
+    schema: assertionRefSha384Schema,
+  },
+  {
+    algo: jwkPubKeyHashAlgorithmSchema.Enum.sha512,
+    schema: assertionRefSha512Schema,
+  },
+];
+
+const getAlgoFromAssertionRef = (
+  assertionRef: string,
+): JwkPubKeyHashAlgorithm => {
+  const found = algoSchemaMap.find(
+    ({ schema }) => schema.safeParse(assertionRef).success,
+  );
+  if (!found)
+    throw new MiddlewareError("Unknown algorithm for given AssertionRef", 403);
+  return found.algo;
+};
+
+const getKeyThumbprintFromSignature = (
+  signatureInput: LollipopSignatureInput,
+): Thumbprint => {
+  const match = /;?keyid="([^"]+)";?/.exec(signatureInput);
+  const thumbprint = match?.[1];
+  const parsed = thumbprintSchema.safeParse(thumbprint);
+  if (!parsed.success)
+    throw new MiddlewareError("Invalid keyid in signature-input", 500);
+  return parsed.data;
+};
+
+const getNonceOrUuidFromSignature = (
+  signatureInput: LollipopSignatureInput,
+): string => {
+  const match = /;?nonce="([^"]+)";?/.exec(signatureInput);
+  return match ? match[1] : ulid();
+};
+
+export const parseLollipopHeaders = async (
+  req: HttpRequest,
+  lollipopClient: LollipopClient,
+): Promise<LollipopHeaders> => {
+  const normalizedHeaders = Object.fromEntries(req.headers.entries());
+  const parsedRequestHeaders =
+    lollipopRequestHeadersSchema.safeParse(normalizedHeaders);
+  if (!parsedRequestHeaders.success)
+    throw new MiddlewareError(`Missing required lollipop headers`, 403);
```

Comment:
```suggestion
    throw new MiddlewareError(`Missing or invalid required lollipop headers`, 403);
```

Here headers could also be invalid?

Reply:
yes, they should satisfy the lollipopRequestHeadersSchema

Reply:
[here](https://github.com/pagopa/io-messages/pull/423/commits/620dbda1f0b12a0ce2829b42db128239f8cd74b4)